### ISSUE TITLE
Blog plugin crashes on Python 3.8 (Netlify)

### ISSUE THREAD

**Issue by funkypenguin** [squidfunk/mkdocs-material#5916]
### Context

I host my mkdocs site on Netlify, which currently uses Python 3.8

### Bug description

When enabling the blog plugin on Python 3.8, `mkdocs build` fails with `TypeError: 'ABCMeta' object is not subscriptable`. 

Full output below:

```
Traceback (most recent call last):
  File "/Users/davidy/.pyenv/versions/mkdocs-blog/bin/mkdocs", line 8, in <module>
    sys.exit(cli())
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/mkdocs/__main__.py", line 283, in build_command
    cfg = config.load_config(**kwargs)
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/mkdocs/config/base.py", line 378, in load_config
    errors, warnings = cfg.validate()
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/mkdocs/config/base.py", line 230, in validate
    run_failed, run_warnings = self._validate()
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/mkdocs/config/base.py", line 188, in _validate
    self[key] = config_option.validate(value)
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/mkdocs/config/config_options.py", line 182, in validate
    return self.run_validation(value)
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/mkdocs/config/config_options.py", line 1064, in run_validation
    self.load_plugin_with_namespace(name, cfg)
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/mkdocs/config/config_options.py", line 1100, in load_plugin_with_namespace
    return (name, self.load_plugin(name, config))
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/mkdocs/config/config_options.py", line 1118, in load_plugin
    plugin_cls = self.installed_plugins[name].load()
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/importlib_metadata/__init__.py", line 209, in load
    module = import_module(match.group('module'))
  File "/Users/davidy/.pyenv/versions/3.8.18/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/material/plugins/blog/plugin.py", line 44, in <module>
    from .structure import Archive, Category, Excerpt, Post, View
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/material/plugins/blog/structure/__init__.py", line 40, in <module>
    from .config import PostConfig
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/material/plugins/blog/structure/config.py", line 24, in <module>
    from .options import PostDate
  File "/Users/davidy/.pyenv/versions/3.8.18/envs/mkdocs-blog/lib/python3.8/site-packages/material/plugins/blog/structure/options.py", line 30, in <module>
    class DateDict(UserDict[str, datetime]):
TypeError: 'ABCMeta' object is not subscriptable
```

### Related links

- [Reporting a bug](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/)
- [Setting up a blog](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/)
- [That time the social plugin broke on Python 3.8](https://github.com/squidfunk/mkdocs-material/issues/5493)

### Reproduction

[9.2.5-python3.zip](https://github.com/squidfunk/mkdocs-material/files/12463815/9.2.5-python3.zip)

### Steps to reproduce

1. Create a fresh python 3.8 env using `pyenv install 3.8`
2. Install mkdocs with `pip install mkdocs-material`
3. Create the absolute minimal site with only the blog plugin enabled (see reproduction)
4. Uncomment the `blog` section under `plugins:` (_the info plugin wouldn't work while it was enabled, since mkdocs would error_)
5. Run `mkdocs build` and observe the error

### Before submitting

- [X] I have read and followed the [bug reporting guidelines](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/).
- [X] I have attached links to [the documentation](https://squidfunk.github.io/mkdocs-material), and possibly related [issues](https://github.com/squidfunk/mkdocs-material/issues) and [discussions](https://github.com/squidfunk/mkdocs-material/discussions).
- [X] I assure that I have [removed all customizations](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/#remove-customizations) before submitting this bug report.
- [X] I have attached a __.zip file__ with a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/).

**Comment by squidfunk** [comment]
Thanks for reporting. Indeed, Python 3.8 should still work. @facelessuser sorry to loop you in, but I'm not sure who to ask otherwise – this should also be resolved by importing `annotations` from `__future__`, as it's related to typings, right?

**Comment by facelessuser** [comment]
@squidfunk, let me take a look

**Comment by squidfunk** [comment]
Thanks for investigating. Hmm. Meh. I'll look into it as soon as I find some time. If we can work around it somehow else, we can also replace it with a another / more compatible solution.

**Comment by facelessuser** [comment]
So, this is my recommendation. Either switch out two implementations:

```
if sys.version_info >= (3, 9):
    class Test(UserDict[str, int]):
        """Class."""
else:
    class Test(UserDict):
        """Class."""
```

Or use some other mapping type that doesn't have this quirk in Python 3.8. I guess you could just not specify the type for all versions of Python as well. `UserDict` typing is kind of broken on 3.8.

**Comment by facelessuser** [comment]
Typing is great when it works, but when it doesn't work, I want to punch it in the face 🙃. This is one of those cases where finding the perfect solution is more work than is worth it. It may be easier to just inherit from `dict` directly.

**Comment by hellt** [comment]
Can't netlify build from a container? This kind of issues won't ever surface (and I strongly believe would make installation way easier) if people would just use the official container image as a default way of building docs and resorting to pypi install as a fallback

**Comment by facelessuser** [comment]
So, this is what I'd probably do with the expectation that in a year, when Python 3.8 is EOL, you can just remove the else. It will maintain parity with what you have, you'll lose some type info in 3.8, but 🤷🏻 . More importantly, it should work on all Python versions. Yeah, ideally you wouldn't have to duplicate things, and if you are willing to rewrite it inheriting from something like `MutableMapping` or `dict`, you could probably replicate the same behavior with a single instance, but you'll have to implement more dictionary methods.

```diff
diff --git a/material/plugins/blog/structure/options.py b/material/plugins/blog/structure/options.py
index d37779185..8d387c5ce 100644
--- a/material/plugins/blog/structure/options.py
+++ b/material/plugins/blog/structure/options.py
@@ -17,7 +17,8 @@
 # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 # FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 # IN THE SOFTWARE.
-
+from __future__ import annotations
+import sys
 from collections import UserDict
 from datetime import date, datetime, time
 from mkdocs.config.base import BaseConfigOption, Config, ValidationError
@@ -27,19 +28,34 @@ from mkdocs.config.base import BaseConfigOption, Config, ValidationError
 # -----------------------------------------------------------------------------
 
 # Date dictionary
-class DateDict(UserDict[str, datetime]):
+if sys.version_info >= (3, 9):
+    class DateDict(UserDict[str, datetime]):
+
+        # Initialize date dictionary
+        def __init__(self, data: dict):
+            super().__init__(data)
 
-    # Initialize date dictionary
-    def __init__(self, data: dict):
-        super().__init__(data)
+            # Initialize date of creation
+            if "created" in data:
+                self.created: datetime = data["created"]
+
+        def __getattr__(self, name: str):
+            if name in self.data:
+                return self.data[name]
+else:
+    class DateDict(UserDict):
+
+        # Initialize date dictionary
+        def __init__(self, data: dict):
+            super().__init__(data)
 
-        # Initialize date of creation
-        if "created" in data:
-            self.created: datetime = data["created"]
+            # Initialize date of creation
+            if "created" in data:
+                self.created: datetime = data["created"]
 
-    def __getattr__(self, name: str):
-        if name in self.data:
-            return self.data[name]
+        def __getattr__(self, name: str):
+            if name in self.data:
+                return self.data[name]
 
 # -----------------------------------------------------------------------------
```

**Comment by facelessuser** [comment]
> Can't netlify build from a container? This kind of issues won't ever surface (and I strongly believe would make installation way easier) if people would just use the official container image as a default way of building docs and resorting to pypi install as a fallback

While running Netlify in a container may be a fine idea, this still needs to be fixed generally as it breaks when used outside a container.

**Comment by hellt** [comment]
> While running Netlify in a container may be a fine idea, this still needs to be fixed generally as it breaks when used outside a container.

Yes, absolutely. I was more alluding to the fact that pypi-based install is [marked](https://squidfunk.github.io/mkdocs-material/getting-started/#with-pip) as _recommended_, while, IMHO, most static-site hosters and CI's would be way shorted in config and would reduce the number of opened issues if they were using container-based installation method.

**Comment by squidfunk** [comment]
Thanks for the input! I'm not a fan of the if/else solution, because if functionality is added, we need to do it in two places. Not great. Can't we just only branch on the base class? Something like:

``` py
BaseDict = UserDict
if sys.version_info >= (3, 9):
    BaseDict = UserDict[str, int]

class DateDict(BaseDict):
    pass
```

Or does that fry the typings? I'm using `UserDict` because MkDocs also uses it as a base for their `Config` class.

**Comment by facelessuser** [comment]
That might work

**Comment by squidfunk** [comment]
Thanks for your expertise and assessment! I'll give it a try.

**Comment by facelessuser** [comment]
I tried it locally, it works

**Comment by funkypenguin** [comment]
> Can't netlify build from a container? This kind of issues won't ever surface (and I strongly believe would make installation way easier) if people would just use the official container image as a default way of building docs and resorting to pypi install as a fallback

I double-checked, it doesn't seem possible to do a Netlify build with an arbitrary container. You have to use one of their (two) available build containers.

**Comment by vedranmiletic** [comment]
> Thanks for reporting. Indeed, Python 3.8 should still work.

> While running Netlify in a container may be a fine idea, this still needs to be fixed generally as it breaks when used outside a container.

Kudos for this attitude in general and supporting such a wide range of versions (3.7+ IIRC), even when it is not convenient or nice. From a user's perspective, it really helps, e.g. some of our machines still run Debian bullseye and FreeBSD with Python 3.9 and it feels good to know that we can continue installing new versions from PyPI for a while and don't have to rush OS upgrades too much.

**Comment by squidfunk** [comment]
Fixed in f2512ded4. Turns out there's a branchless solution: using `typing.Dict`.

I've tested the fix in Python 3.7, 3.8 and 3.11 – all good.

**Comment by facelessuser** [comment]
Yep `typing.Dict` is even better. I didn't know if it gave you access to `self.data`.

**Comment by squidfunk** [comment]
You can just use `self` ☺️

**Comment by squidfunk** [comment]
Released as part of [9.2.6](https://squidfunk.github.io/mkdocs-material/changelog/#9.2.6).
